### PR TITLE
add support for triggering the stellar payment confirm screen in the UI CORE-9471

### PR DESCRIPTION
**Pull request by mmaxim**
**Go**
- Change the Stellar UI flow to better match the actual UI. Add a `ChatPostReadyToSend` callback from `PostTextNonblock` in order for the UI to know when to add the message to the local store.
- Refactor `wallet.Sender` to move more `DescribePayment` logic into itself.
- Optimize how we obtain fullnames and participant lists. 
- Add tests

**JS**
@keybase/react-hackers 
- Add new component and stories for the in-chat send confirm screen.
- Change the `messageSend` send flow to use the new stellar UI flow. Notably, the call into `PostTextNonblock` is now a Saga RPC, and runs the callbacks to drive the confirm screen flow. I've tested this a lot and see no performance degradation on normal sends.

image

### CODE REVIEW

**Comment by mmaxim** [comment]
image
Example

**Comment by chrisnojima** [comment]
There are a couple of aspects of the sagas we should discuss tomorrow morning

File: shared/chat/confirm-payment/index.js [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+
+const styles = Styles.styleSheetCreate({
+  buttonContainer: Styles.platformStyles({
+    common: {
+      justifyContent: 'space-between',
+    },
+  }),
+  cancelButton: Styles.platformStyles({
+    isElectron: {
+      height: 40,
```
Comment (chrisnojima) [inline shared/chat/confirm-payment/index.js]: we have sizes in global margins. this is 'large'

File: shared/chat/confirm-payment/index.js [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+  errorText: {
+    color: Styles.globalColors.red,
+  },
+  fullErrorContainer: Styles.platformStyles({
+    isElectron: {
+      padding: 20,
+    },
+  }),
+  headerText: Styles.platformStyles({
+    common: {
```
Comment (chrisnojima) [inline shared/chat/confirm-payment/index.js]: why not use bodytiny

File: shared/chat/confirm-payment/index.js [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+  paymentContainer: Styles.platformStyles({
+    common: {
+      borderColor: Styles.globalColors.lightGrey,
+      borderStyle: 'solid',
+      borderWidth: 1,
+      justifyContent: 'space-between',
+      padding: Styles.globalMargins.tiny,
+    },
+    isElectron: {
+      margin: -1,
```
Comment (chrisnojima) [inline shared/chat/confirm-payment/index.js]: why is this needed

File: shared/chat/confirm-payment/index.js [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+  }),
+  paymentsContainer: Styles.platformStyles({
+    isElectron: {
+      minHeight: 150,
+    },
+  }),
+  submitButton: Styles.platformStyles({
+    isElectron: {
+      height: 40,
+      minWidth: 225,
```
Comment (chrisnojima) [inline shared/chat/confirm-payment/index.js]: why is this needed

File: shared/chat/confirm-payment/index.js [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+      paddingRight: 8,
+    },
+  }),
+  totalContainer: Styles.platformStyles({
+    common: {
+      alignItems: 'center',
+      backgroundColor: Styles.globalColors.purple,
+    },
+    isElectron: {
+      height: 242,
```
Comment (chrisnojima) [inline shared/chat/confirm-payment/index.js]: why is this needed

File: shared/chat/confirm-payment/index.js [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    },
+  }),
+  xlmTotal: Styles.platformStyles({
+    common: {
+      color: Styles.globalColors.white,
+    },
+    isElectron: {
+      fontSize: 24,
```
Comment (chrisnojima) [inline shared/chat/confirm-payment/index.js]: why not header big

File: shared/constants/types/chat2/index.js [hunk 7]
```diff
@@ -27,6 +27,13 @@ export type _QuoteInfo = {
 }
 export type QuoteInfo = I.RecordOf<_QuoteInfo>
 
+export type PaymentConfirmInfo = {
+  error?: string,
+  summary?: RPCChatTypes.UIChatPaymentSummary,
+  // $FlowIssue let me use any here
+  response: any,
```
Comment (chrisnojima) [inline shared/constants/types/chat2/index.js]: we don't want this on purpose

File: shared/actions/chat2/index.js [hunk 8]
```diff
@@ -1271,64 +1271,106 @@ const messageRetry = (action: Chat2Gen.MessageRetryPayload, state: TypedState) =
...
-      context: {type: 'sent'},
-      messages: [newMsg],
+    const addMessage = (response: any) => [
+      Saga.put(
+        Chat2Gen.createMessagesAdd({
+          context: {type: 'sent'},
+          messages: [newMsg],
+        })
+      ),
+      Saga.callUntyped(function() {
```
[the remark this answers is not in the document]
Comment (mmaxim) [reply shared/actions/chat2/index.js]: Discussed offline

File: shared/chat/confirm-payment/index.js (hunk 3 above)
Comment (mmaxim) [reply shared/chat/confirm-payment/index.js]: Keep the border from doubling up.

File: shared/chat/confirm-payment/index.js (hunk 4 above)
Comment (mmaxim) [reply shared/chat/confirm-payment/index.js]: So the button is the same width regardless of XLM amount

**Comment by mmaxim** [comment]
@chrisnojima PTAL

**Comment by mmaxim** [comment]
Actually now, sorry.

File: shared/actions/chat2/index.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/actions/chat2/index.js]: Let's leave a comment here

File: shared/actions/chat2/index.js [hunk 9]
```diff
@@ -1271,64 +1271,108 @@ const messageRetry = (action: Chat2Gen.MessageRetryPayload, state: TypedState) =
...
+        response && response.result()
+      }),
+    ]
+    const onShowConfirm = () => [
+      Saga.put(Chat2Gen.createSetPaymentConfirmInfo({info: null})),
+      Saga.put(
+        RouteTreeGen.createNavigateAppend({
+          path: [
+            {
+              props: {},
```
Comment (chrisnojima) [inline shared/actions/chat2/index.js]: Just pass the name. Don't need props or selected

File: shared/actions/chat2/index.js [hunk 10]
```diff
@@ -1271,64 +1271,108 @@ const messageRetry = (action: Chat2Gen.MessageRetryPayload, state: TypedState) =
...
+            },
+          ],
+        })
+      ),
+    ]
+    const onDataConfirm = ({summary}, response) => {
+      stellarConfirmWindowResponse = response
+      return Saga.put(
+        Chat2Gen.createSetPaymentConfirmInfo({
+          info: {
```
Comment (chrisnojima) [inline shared/actions/chat2/index.js]: Why have info

File: shared/actions/chat2/index.js [hunk 11]
```diff
@@ -1271,64 +1271,108 @@ const messageRetry = (action: Chat2Gen.MessageRetryPayload, state: TypedState) =
...
+          },
+        })
+      )
+    }
+    const onDataError = ({message}, response) => {
+      stellarConfirmWindowResponse = response
+      return Saga.put(
+        Chat2Gen.createSetPaymentConfirmInfo({
+          info: {
+            error: message,
```
Comment (chrisnojima) [inline shared/actions/chat2/index.js]: If calls have errors there's a way to say this in json. See canError

File: shared/actions/chat2/index.js [hunk 12]
```diff
@@ -1271,64 +1271,108 @@ const messageRetry = (action: Chat2Gen.MessageRetryPayload, state: TypedState) =
...
+      )
+    }
+    yield RPCChatTypes.localPostTextNonblockRpcSaga({
+      customResponseIncomingCallMap: {
+        'chat.1.chatUi.chatPostReadyToSend': addMessage,
+        'chat.1.chatUi.chatStellarDataConfirm': onDataConfirm,
+        'chat.1.chatUi.chatStellarDataError': onDataError,
+      },
+      incomingCallMap: {
+        'chat.1.chatUi.chatStellarDone': p => Saga.put(navigateUp()),
```
Comment (chrisnojima) [inline shared/actions/chat2/index.js]: Likely only want to do this if the modal is up

File: shared/actions/chat2/index.js [hunk 13]
```diff
@@ -1338,6 +1382,12 @@ const messageSendWithError = (result, action) => {
   logger.info('[MessageSend] error')
 }
 
+let stellarConfirmWindowResponse: any = null
+
+const confirmScreenResponse = (state: TypedState, action: Chat2Gen.ConfirmScreenResponsePayload) => {
+  stellarConfirmWindowResponse && stellarConfirmWindowResponse.result(action.payload.accept)
```
Comment (chrisnojima) [inline shared/actions/chat2/index.js]: Should clear this also

File: shared/chat/confirm-payment/index.js (hunk 4 above)
Comment (chrisnojima) [reply shared/chat/confirm-payment/index.js]: Why not flex 1

**Comment by mmaxim** [comment]
Ok @chrisnojima, PTAL again.

File: shared/actions/chat2/index.js [hunk 14]
```diff
@@ -1271,64 +1271,103 @@ const messageRetry = (action: Chat2Gen.MessageRetryPayload, state: TypedState) =
...
 
-  logger.info('[MessageSend]', 'non-empty text?', text.stringValue().length > 0)
+    const addMessage = (p, response) => [
+      Saga.put(
+        Chat2Gen.createMessagesAdd({
+          context: {type: 'sent'},
+          messages: [newMsg],
+        })
+      ),
+      Saga.callUntyped(function() {
```
Comment (chrisnojima) [inline shared/actions/chat2/index.js]: lets add a comment here

File: shared/actions/chat2/index.js [hunk 15]
```diff
@@ -1271,64 +1271,103 @@ const messageRetry = (action: Chat2Gen.MessageRetryPayload, state: TypedState) =
...
+        response && response.result()
+      }),
+    ]
+    const onShowConfirm = () => {
+      stellarConfirmPaymentModalShown = true
+      return [
```
Comment (chrisnojima) [inline shared/actions/chat2/index.js]: I believe this will show a warning. Should be Saga.sequential()

File: shared/actions/chat2/index.js [hunk 16]
```diff
@@ -1271,64 +1271,103 @@ const messageRetry = (action: Chat2Gen.MessageRetryPayload, state: TypedState) =
...
+            path: ['paymentsConfirm'],
+          })
+        ),
+      ]
+    }
+    const onHideConfirm = () => {
+      if (!stellarConfirmPaymentModalShown) {
+        return
+      }
+      stellarConfirmPaymentModalShown = false
```
Comment (chrisnojima) [inline shared/actions/chat2/index.js]: i actually meant checking the routeState. when this modal is up you can't do anything else? i'd lean towards removing this boolean. seems error prone

File: shared/actions/chat2/index.js [hunk 17]
```diff
@@ -1338,6 +1377,13 @@ const messageSendWithError = (result, action) => {
   logger.info('[MessageSend] error')
 }
 
+let stellarConfirmWindowResponse: any = null
```
Comment (chrisnojima) [inline shared/actions/chat2/index.js]: can be typed

File: shared/actions/chat2/index.js [hunk 18]
```diff
@@ -1271,64 +1271,103 @@ const messageRetry = (action: Chat2Gen.MessageRetryPayload, state: TypedState) =
...
-  //
-  // We put the addMessage on the back in case the service provides chat thread data in between the
-  // addMessage and postText action. upgradeMessage should be a no-op in the case that the message
-  // that is in the store on the outboxID has been sent.
-  return Saga.sequentially([addMessage, postText, addMessage])
-}
+    // Do some logging to track down the root cause of a bug causing
+    // messages to not send. Do this after creating the objects above to
+    // narrow down the places where the action can possibly stop.
+    logger.info('[MessageSend]', 'non-empty text?', text.stringValue().length > 0)
```
Comment (chrisnojima) [inline shared/actions/chat2/index.js]: likely want to clear stellarConfirmWindowResponse here too

Comment (mmaxim) [reply shared/actions/chat2/index.js]: It gets cleared right before we show it.

Comment (chrisnojima) [reply shared/actions/chat2/index.js]: but better to not let it lying around right. seems more error prone

File: shared/actions/chat2/index.js (hunk 15 above)
Comment (mmaxim) [reply shared/actions/chat2/index.js]: Doesn't work here, using Saga.all
